From patchwork Fri Jul 27 16:18:50 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Jim Mattson <jmattson@google.com>
X-Patchwork-Id: 10547415
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 8A2C1174A
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri, 27 Jul 2018 16:19:02 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 7A1552C14B
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri, 27 Jul 2018 16:19:02 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 6DB592C159; Fri, 27 Jul 2018 16:19:02 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-15.5 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI,
	USER_IN_DEF_DKIM_WL autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 2582F2C14B
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri, 27 Jul 2018 16:19:02 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S2388585AbeG0Rlg (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Fri, 27 Jul 2018 13:41:36 -0400
Received: from mail-yw0-f201.google.com ([209.85.161.201]:38929 "EHLO
        mail-yw0-f201.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1730205AbeG0Rlg (ORCPT <rfc822;kvm@vger.kernel.org>);
        Fri, 27 Jul 2018 13:41:36 -0400
Received: by mail-yw0-f201.google.com with SMTP id u1-v6so2845166ywg.6
        for <kvm@vger.kernel.org>; Fri, 27 Jul 2018 09:19:00 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=google.com; s=20161025;
        h=date:message-id:mime-version:subject:from:to:cc;
        bh=G3NVisfEHK4P9wj0G1q02EdLXEHEAs82HaWC6DFgRYY=;
        b=c0ctA1lMeBMfA8Vp6ppMaQAuD85n6oHXEZZ6HMLleKMHFJvadEnkcBIo/KaXNgLZI8
         nTGnqzp9y2qBOI1hlV86wSds1Qc+4q1uVkyFmOllbuQdTVOfZzwjWr/5bkWi6D5vcnDt
         Z0NIXMdlnc++2E4xTqFpsu4vEbcutFd3zNZ2He+8hL2zFid58QOhyI1ozZtTJGTQkbig
         Qpez3OL6/0Xzcnet4Ttmovsyaje2IGpg+IXyvq9J3kclheQNVqUKWi4eYZSK6kAAwX1J
         Fs4LGluxdYQIC2X2DEEdQR7Xnf15SSxZN0WqoX2mtkhUw9PWPW/J+iGTplH8NvUf/6VK
         W+5g==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:date:message-id:mime-version:subject:from:to:cc;
        bh=G3NVisfEHK4P9wj0G1q02EdLXEHEAs82HaWC6DFgRYY=;
        b=czbHRnbBvpOJ5lrqy0DjOgVXvzQVS/SXccDFZPjnF5YFAKj99AeIkd/DYbx9WgUUZl
         lJflnbDmrV04S4tNQ/2VWU54mMUPNmGZgwW9ng5aRDuGSmwCDIwQtbmcSoMyS2cCEzNw
         EyFKxrFZ+w/oKDQabV77VgEhHtTABHijJN87gaitGFDflpSWwWqwlShcgY3c+xMFgDsv
         BkKG5xcPdF3QitmNp/TnhOwiYNkg5ezXO9ePVe5V3JrKbYPdlQSeVcfLi+nXDgo1LuwB
         QK8WwyUtNAR2fEmewEXjy4D7JN6qBBeA+eVkpBlfxQrnCbtBeQSexjsjnwWbWQ3SrxlO
         GSVg==
X-Gm-Message-State: AOUpUlEOgYtRZl1dxvJ9k/rOvAfbArECDWa5+R7GhsAzhqYxkZSfV0HA
        UgySvD/pJdvneQZtp834KHD2l3GJ1iWKcRv4BcyNZZyCk8xYqhuPCjjsrZwFhq9ynNQR0adZvIe
        WjiF9XILFuQKb+kA8UkYc5bm1Il9aThiRxXezl5vyFbky4p4xgTjtVWeltteVczc=
X-Google-Smtp-Source: 
 AAOMgpcEq5r8EYKpASog7KHmLYMz32qHtIEhsZab3prtK+uIgT4hAfTUX34ABgD0JzFi9vatxWx7K9Yn5qx7UQ==
X-Received: by 2002:a81:8602:: with SMTP id
 w2-v6mr2110400ywf.77.1532708339277;
 Fri, 27 Jul 2018 09:18:59 -0700 (PDT)
Date: Fri, 27 Jul 2018 09:18:50 -0700
Message-Id: <20180727161850.172739-1-jmattson@google.com>
Mime-Version: 1.0
X-Mailer: git-send-email 2.18.0.345.g5c9ce644c3-goog
Subject: [PATCH] kvm: nVMX: Fix fault vector for VMX operation at CPL > 0
From: Jim Mattson <jmattson@google.com>
To: kvm@vger.kernel.org, Felix Wilhelm <fwilhelm@google.com>,
        Paolo Bonzini <pbonzini@redhat.com>
Cc: Jim Mattson <jmattson@google.com>
Content-Type: text/plain; charset="UTF-8"
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

The fault that should be raised for a privilege level violation is #GP
rather than #UD.

Fixes: 727ba748e110b4 ("kvm: nVMX: Enforce cpl=0 for VMX instructions")
Signed-off-by: Jim Mattson <jmattson@google.com>
Reviewed-by: Paolo Bonzini <pbonzini@redhat.com>
Reviewed-by: David Hildenbrand <david@redhat.com>
---
 arch/x86/kvm/vmx.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/arch/x86/kvm/vmx.c b/arch/x86/kvm/vmx.c
index 5d8e317c2b04f..cb51808c73206 100644
--- a/arch/x86/kvm/vmx.c
+++ b/arch/x86/kvm/vmx.c
@@ -7941,7 +7941,7 @@ static int handle_vmon(struct kvm_vcpu *vcpu)
 
 	/* CPL=0 must be checked manually. */
 	if (vmx_get_cpl(vcpu)) {
-		kvm_queue_exception(vcpu, UD_VECTOR);
+		kvm_inject_gp(vcpu, 0);
 		return 1;
 	}
 
@@ -8005,7 +8005,7 @@ static int handle_vmon(struct kvm_vcpu *vcpu)
 static int nested_vmx_check_permission(struct kvm_vcpu *vcpu)
 {
 	if (vmx_get_cpl(vcpu)) {
-		kvm_queue_exception(vcpu, UD_VECTOR);
+		kvm_inject_gp(vcpu, 0);
 		return 0;
 	}
 
